files/class-vip-filesystem.php

```diff
@@ -129,4 +145,182 @@ public function filter_upload_dir( $params ) {
+		$filename = sanitize_file_name( $filename );
+
+		// Setting `ext` and `type` to empty will fail the upload because Go doesn't allow unfiltered uploads
+		// See `_wp_handle_upload()`
+		if ( ! $this->check_filetype_with_backend( $filename ) ) {
+			$filetype_data['ext']             = '';
+			$filetype_data['type']            = '';
+			// Never set this true, which leaves filename changing to dedicated methods in this class
+			$filetype_data['proper_filename'] = false;
+		}
+
+		return $filetype_data;
+	}
+
+	/**
+	 * Use the VIP Filesystem API to check for filename uniqueness
+	 *
+	 * The `unique_filename` API will return an error if file type is not supported
+	 * by the VIP Filesystem.
+	 *
+	 * @since   1.0.0
+	 * @access  protected
+	 *
+	 * @param   string      $filename
+	 *
+	 * @return  bool        True if filetype is supported. Else false
+	 */
+	protected function check_filetype_with_backend( $filename ) {
+		$upload_path = $this->get_upload_path();
+
+		$file_path = $upload_path . $filename;
+
+		$result = $this->stream_wrapper->client->get_unique_filename( $file_path );
+
+		if ( is_wp_error( $result ) ) {
+			if ( 'invalid-file-type' !== $result->get_error_code() ) {
+				trigger_error( $result->get_error_message(), E_USER_WARNING );
+			}
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Return uploads path
+	 *
+	 * Different from wp_uploads_dir() as that will return the `vip://` protocol too
+	 *
+	 * @since   1.0.0
+	 * @access  private
+	 *
+	 * @return  string The uploads path
+	 */
+	private function get_upload_path() {
+		$upload_dir_path = wp_get_upload_dir()['path'];
+		return substr( $upload_dir_path, strlen( self::PROTOCOL . '://' ) );
+	}
+
+	/**
+	 * Deletes the file and purge the cache
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @param   string $file_path
+	 *
+	 * @return  string
+	 */
+	public function filter_delete_file( $file_path ) {
+		// To ensure we don't needlessly fire off deletes for all sizes of the same image, of
+		// which all except the first result in 404's, keep accounting of what we've deleted.
+		static $deleted_uris = array();
+
+		$file_path = $this->clean_file_path( $file_path );
+
+		$file_uri  = $this->get_file_uri_path( $file_path );
+
+		if ( in_array( $file_uri, $deleted_uris, true ) ) {
+			// This file has already been successfully deleted from the file service in this request
+			return '';
+		}
+
+		if ( ! unlink( $file_path ) ) {
+			return '';
+		}
+
+		// Set our static so we can later recall that this file has already been deleted and purged
+		$deleted_uris[] = $file_uri;
+
+		// We successfully deleted the file, purge the file from the caches
+		$this->purge_file_cache( $file_uri );
+
+		// Return empty value so that `wp_delete_file()` won't try to `unlink` again
+		return '';
+	}
+
+	/**
+	 * Remove duplicate uploads base directory from file path
+	 *
+	 * Some of the intermediate file paths have the uploads `basedir` occur twice so we will need to
```

Comment:
Do we know why the basedir is showing up twice? That sounds like a bug somewhere.

Reply:
Ah, it's probably the path_join (https://core.trac.wordpress.org/browser/trunk/src/wp-includes/post.php#L5483), which doesn't handle URIs with protocols: https://developer.wordpress.org/reference/functions/path_join/